### DISCUSSION TITLE
Resolving URIs to handlers

### DISCUSSION THREAD

**Discussion by sazzer** [actix/actix-web#2559] [Q&A] [answered]
Hi there,

The existence of `HttpRequest::url_for` is fantastic for building real REST APIs. I can use this to generate URIs for other resources so that I can correctly generate the hypermedia graph.

However, what I can't quite work out how to do is the opposite. To take a URL that was was provided to me and convert it back into the details of what it resolves to.

If I understand it right, I can use `HttpRequest::resource_map().match_name()` to get the *name* of the handler that matches a given URL, but this doesn't give me the binds.

For example, if I had:
```
    config.service(
        web::resource("/users/{id}")
            .name("users_user_by_id")
            .route(web::get().to(get::handler)),
    );
```

Then I could use `req.resource_map().match_name("http://localhost:8000/users/abc")` to get `Some("users_user_by_id")`, but that would *not* tell me that the `{id}` bind was assigned the value `"abc"`.

Is there any way to get this full expansion?

Cheers

**Comment by robjtede** [comment]
If you mean the path dynamic segments within the current handler, than use the [Path](https://actix.rs/actix-web/actix_web/web/struct.Path.html) extractor.

**Reply by sazzer** [reply]
Unless I'm missing something though, that's used to parse the *current* request. Not an arbitrary URI that might be passed to another request.

For example, in a hypermedia API I might make a request like: `GET /posts?user=http://blog.example.com/users/abc` - i.e. a call to `/posts` with a query parameter that is the URI to a user.

What this means is that the handler for `web::resource("/posts").route(web::get().to(search::handler))` needs to be able to:
* Take the string `http://blog.example.com/users/abc`.
* Parse it as a URI.
* Get an indication that this URI matches the route `users_user_by_id`.
* Get the string `"abc"` as the value for the `{id}` bind.

That last bullet point is the one that I can't work out how to do.

**Reply by robjtede** [reply]
Ahh I see what you mean. Doesn't look like this is possible with current methods.

Tracking this feature here: https://github.com/actix/actix-web/issues/2571

**Comment by robjtede** [comment, marked as answer]
Not currently possible, feature is being tracked: https://github.com/actix/actix-web/issues/2571
